Use 'Catch' for unit testing

kevinushey opened the pull request:
NOTE: we can leave this for after Preview-2 merge, but we might want to merge soon to avoid inevitable merge conflicts due to the `session` -> `rsession` rename (see more later)

This PR adds some infrastructure for using [Catch](https://github.com/philsquared/Catch) to run unit tests in our C++ code. I provide a couple of small abstractions over Catch in `TestThat.hpp` and `TestRunner.hpp`.

We use a very similar interface to the `testthat` package for running tests. Any implementation file that wants to add unit tests can simply include the `<tests/TestThat.hpp>` header, and declare a test block with e.g.

```cpp
test_that("this function works") {
    expect_true(<functionCall>);
}
```

`test_that` and `expect_*` are macros that map to the Catch macro counterparts; we might want to use different names to avoid macro collisions just in case but this seems to work fine (and we don't use snake casing anywhere else in the code base)

Unfortunately, on OS X, catch includes `<sys/proc.h>`, which has:

```
struct session;
```

so one of the side effects is that I had to rename the `session` namespace to `rsession`. (That's all the noise you see in the `.cpp` files)

The `test-runner` CMake project is set up to glop up any file matching the glob `*Tests.cpp`, so adding new tests for a file is as simple as creating a new file `<File>Tests.cpp` and writing the Catch tests within.

My `cmake`-fu is a bit lacking so let me know if anything I've done seems particularly weird / off!

jjallaire commented:
Excited to add this to the project! (Catch looks excellent). Let's talk realtime about the topology of the test project and how subprojects get their tests included (I don't currently see a way for binaries like rserver and rsession to define tests but I could be missing something -- again, let's discuss realtime on Tuesday).

In terms of the namespace issue, I think the answer is the (more annoying to implement) addition of an "rstudio" namespace outside all of the existing top level namespaces (core, session, server, etc.). This is actually what I also suggested for this PR (which ran into the problem that Boost 1.57 defines a boost::core namespace):

https://github.com/rstudio/rstudio/pull/218

kevinushey commented:
Oops, you're right -- perhaps my first approach would indeed be the better one (conditionally compile testing into each of the executables, rather than having a separate test executable).

kevinushey commented:
I went ahead with a slightly different approach -- library targets now also build an executable for Debug builds, which we run on a `make all`; executables gain the `--test` command line argument (currently only for `rsession`) when built in Debug mode.

This ensures that tests don't get compiled into the release (to avoid bloating the executables / libraries).

jjallaire commented on src/cpp/core/CMakeLists.txt:
```diff
@@ -282,3 +281,35 @@ target_link_libraries(
+      ${CORE_SYSTEM_LIBRARIES}
+   )
+   
+   # run tests after build
```
I don't think we should run the tests after every build as this will quickly limit us to tests that don't run for more than (in total) ~ 3 seconds. We can use the --test argument along with a script like we do for other binaries (e.g. ./rserver-dev or ./rstudio-dev).

kevinushey replied:
Hmm, good point. It would be best if we could automate the running / reporting of unit tests though -- maybe we can talk a bit about that tomorrow?

jjallaire replied:
It might just be as simple as a convention that you always run the "smoke tests" before you check in. Alternatively we could have a post-commit hook somewhere else that builds and runs the tests.

jjallaire replied:
Let's discuss when you get in today.

jjallaire commented on src/cpp/CMakeLists.txt:
```diff
@@ -184,6 +186,9 @@ endif()
 add_subdirectory(core)
 
+# test executable
+add_subdirectory(tests/cpp)
```
I think we should use the RSTUDIO_UNIT_TESTS_ENABLED around this call rather than inside the tests/cpp/CMakeLists.txt

kevinushey replied:
Good point, thanks!

kevinushey commented:
Bleh. Even with the new `rstudio` namespace, we still get loads of conflicts from the `session` struct brought in by `Catch`. (We don't get saved because `session::foo` is ambiguous as to whether we want `foo` from the `rstudio::session` namespace or the global `session` object leaked in)

Some workarounds:

1. Rename our `session` to `rsession`,
2. Have `namespace rsession = rstudio::session` in any test files / runners,
3.

kevinushey commented:
Okay, tidied up by using `rsession` as an alias in `SessionMain.cpp`.

Only thing left re: this PR is automating tests, but not running them after every build.

kevinushey commented:
I disabled the automatic test running -- I think this should be okay for merge if we want. We can think about automated running of tests (e.g. through commit hooks) later on?

jjallaire commented:
I didn't see the --test command line options, did I miss them?

On Wed, Jan 21, 2015 at 4:55 PM, Kevin Ushey <notifications@github.com>
wrote:

> I disabled the automatic test running -- I think this should be okay for
> merge if we want. We can think about automated running of tests (e.g.
> through commit hooks) later on?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/rstudio/rstudio/pull/275#issuecomment-70929629>.
>

kevinushey commented:
The `--test` command line option is specified on the `rsession` executable -- in `TestRunner.hpp`:

    https://github.com/rstudio/rstudio/blob/feature/catch-unit-testing/src/cpp/tests/cpp/tests/TestRunner.hpp#L34-L56

which is called in `SessionMain.cpp`:

    https://github.com/rstudio/rstudio/blob/feature/catch-unit-testing/src/cpp/session/SessionMain.cpp#L2853-L2855

For `core`, we simply have a `TestMain.cpp` that looks like this:

    https://github.com/rstudio/rstudio/blob/feature/catch-unit-testing/src/cpp/core/TestMain.cpp

which is simple enough to ensure that an executable is generated that runs Catch unit tests.

kevinushey commented:
If we merge, let me double check that this branch builds on Linux + Windows first.

jjallaire commented on src/cpp/core/CMakeLists.txt:
```diff
@@ -282,3 +281,29 @@ target_link_libraries(
+
+  file(GLOB_RECURSE CORE_TEST_FILES "*Tests.cpp")
+
+  add_executable(rstudio-core-bin
```
Let's call this rstudio-core-tests

kevinushey commented:
@jjallaire, I've re-organized things and I think this should be good to go now. I also tested a build locally on Windows and Ubuntu and all worked on my side.

I added an option, `--run-tests`, using the boost program options framework. Aside -- I noticed that our validator here:

https://github.com/rstudio/rstudio/blob/master/src/cpp/core/ProgramOptions.cpp#L38-L58

doesn't accept simple boolean flag arguments (those don't enter the `variables_map`). Is that intended? I worked around it by setting both default and implicit values for the `--run-tests` flag (so it behaves like a flag but still enters the map).

I also re-organized the code a bit so that the `cpp/tests` directory is not considered a system directory (this was kind of a hack; and had the unfortunate side effect of not tracking changes within that directory). Instead, it's a CMake 'global' variable (cached variable) that's used across the various sub-projects for includes.

I also added override directories into the `rsession-dev.conf` so that running `rsession` directly can succeed.

I also simplified the `TestRunner.hpp` script; it's still necessary to provide the `catch.hpp` header and the `run()` command, but it's now called after R's initialization has finished.

Finally, the tests are all run by the `rstudio-tests` shell script, which calls the various test executables (with appropriate command line arguments) one by one.

Let me know if you like how things look.

jjallaire commented on src/cpp/conf/rsession-dev.conf:
```diff
@@ -27,6 +27,10 @@ r-resources-path=${CMAKE_CURRENT_SOURCE_DIR}/session/resources
+# override home and doc dirs
+r-home-dir-override=${LIBR_HOME}
+r-doc-dir-override=${LIBR_DOC_DIR}
+
```
The only issue with doing this is that it makes it hard for us to test features of rserver that dynamically change the version of R being run (e.g. the user profile feature in rstudio server pro). There is another conf/-dev pair called "rdesktop-dev" that I believe does what you want (and contains the -override flags already).

jjallaire commented on src/cpp/rstudio-tests.in:
```diff
@@ -0,0 +1,9 @@
+echo Running 'rsession' tests...
+${CMAKE_CURRENT_BINARY_DIR}/session/rsession \
+    --run-tests \
+    --config-file=${CMAKE_CURRENT_BINARY_DIR}/conf/rsession-dev.conf
```
I think if this just becomes rdesktop-dev.conf then we'll be good to go

jjallaire commented on src/cpp/session/SessionMain.cpp:
```diff
@@ -2976,10 +2990,13 @@ int main (int argc, char * const argv[])
-      error = startHttpConnectionListener();
-      if (error)
-         return sessionExitFailure(error, ERROR_LOCATION);
+      // start http connection listener (TODO: not if running tests)
```
Why not start the connection listener when we are running tests? (might as well test the fact that it can initialize, bind to it's port, etc.?)

jjallaire commented on src/cpp/session/SessionOptions.cpp:
```diff
@@ -82,12 +82,19 @@ core::ProgramStatus Options::read(int argc, char * const argv[])
+   options_description runTests("tests");
+   runTests.add_options()
+         (kRunTestsSessionOption,
+          value<bool>(&runTests_)->default_value(false)->implicit_value(true),
```
This is indeed the correct way to handle an option that has no value.

jjallaire commented:
The PR looks great save for those couple of comments I made. Let me know when those tweaks are in and I'll merge. Thrilled to have this be part of our system!

kevinushey commented:
Alright, I modified the PR re: your comments and think this is good to go. (I just tested a build on my Windows VM to make sure it builds okay there and all looks well there too)

kevinushey commented:
Hm, there are some failures on Linux. Let me check these out.

kevinushey commented:
Nevermind, it looks okay -- I think I had forgotten to re-run cmake.
